fix: seperate nodes entry for backward compatibility

yue4u opened the pull request:
This PR attempts to address compatibility issues with three before r161.

Removed `export * from './nodes'` in `three-vrm-materials-mtoon` and created seperated `./nodes` entry in both `three-vrm-materials-mtoon` and `three-vrm`

yue4u committed:
fix: seperate nodes entry for backward compatibility

yue4u commented on tsconfig.json:
```diff
@@ -3,7 +3,7 @@
     "module": "es6",
     "lib": ["es2018", "dom"],
-    "moduleResolution": "node",
+    "moduleResolution": "Bundler",
```
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-0.html#--moduleresolution-bundler
Expect this not to affect end user

0b5vr replied:
why is the first letter capitalized though?

yue4u replied:
From autocompletion. (guess this is the default)
![Screenshot 2024-06-28 at 13 32 36](https://github.com/pixiv/three-vrm/assets/26110087/a4af5eca-328d-4737-b2a9-fdef5d0ad188)

0b5vr replied:
huh. It's kinda weird for me but it's ok.

yue4u commented on packages/three-vrm/src/nodes/index.ts:
```diff
@@ -0,0 +1 @@
+export * from '@pixiv/three-vrm-materials-mtoon/nodes';
```
Should we add some three version checks and log info here or inside `@pixiv/three-vrm-materials-mtoon/nodes`? IMO it's not necessary

0b5vr replied:
we already have a warning inside `MToonNodeMaterial` so I think it's not needed

yue4u replied:
I think MToonNodeMaterial'constructor will never be executed if `three/addons/nodes/Nodes` not exist?

0b5vr replied:
Ahh, yes it's possible to cause an error before `MToonNodeMaterial` instantiates.

Does this help?

![image](https://github.com/pixiv/three-vrm/assets/7824814/dea1f584-4f4d-4adc-8056-c737688bafa4)

yue4u commented on README.md:
````diff
@@ -92,7 +92,7 @@ Code like this:
 ```javascript
 import * as THREE from 'three';
-import { GLTFLoader } from 'three/examples/jsm/loaders/GLTFLoader';
+import { GLTFLoader } from 'three/examples/jsm/loaders/GLTFLoader.js';
````
align to three export

yue4u commented on bin/build.mjs:
```diff
@@ -55,16 +55,46 @@ async function buildPackage(absWorkingDir) {
+    };
+    let outFilename = addSuffix(filename);
+
+    const extraEntryPoints = ['@pixiv/three-vrm', '@pixiv/three-vrm-materials-mtoon'].includes(packageJson.name)
```
doing this here to modify cjs,esm outputs altogether

0b5vr replied:
This section looks surprisingly cryptic compared to what it actually does 😩

yue4u commented on bin/build.mjs:
```diff
@@ -55,16 +55,46 @@ async function buildPackage(absWorkingDir) {
     };
   };
 
+  /**
```
Coundn't find a better way to do this. IMO it's too wasteful to bundle shaders and everything from three-vrm-materials-mtoon main entry again

0b5vr replied:
I will find a way to quit `MToonNodeMaterialLoaderPlugin`. Maybe having a factory pattern or something would help this

yue4u replied:
removed in 6d40b26a5d52d82c00d40b66f72874e5c507744b

0b5vr commented on bin/build.mjs:
```diff
@@ -55,16 +55,46 @@ async function buildPackage(absWorkingDir) {
+    };
+    let outFilename = addSuffix(filename);
+
+    const extraEntryPoints = ['@pixiv/three-vrm', '@pixiv/three-vrm-materials-mtoon'].includes(packageJson.name)
+      ? { [addSuffix('nodes/index')]: 'src/nodes/index.ts' }
+      : {};
```
```suggestion
    const needsNodesBuild = ['@pixiv/three-vrm', '@pixiv/three-vrm-materials-mtoon'].includes(packageJson.name);
    const extraEntryPoints = needsNodesBuild
      ? { [addSuffix('nodes/index')]: 'src/nodes/index.ts' }
      : {};
```

0b5vr commented on bin/build.mjs:
```diff
@@ -55,16 +55,46 @@ async function buildPackage(absWorkingDir) {
   const entryPoints = (format) => {
-    let outFilename = filename + (format === 'esm' ? '.module' : '');
-    outFilename += DEV ? '' : '.min';
+    /**
+     * @param {string} base
+     */
```
```suggestion
    /**
     * Add suffix according to the format to the given basename of the output file.
     * @param {string} base Basename of the output file
     * @returns {string} `<base>[.module][.min]`
     */
```

0b5vr commented on bin/build.mjs:
```diff
@@ -55,16 +55,46 @@ async function buildPackage(absWorkingDir) {
+
     return {
       entryPoints: {
         [outFilename]: 'src/index.ts',
```
I think we can directly put `addSuffix(filename)` here as `extraEntryPoints` does

yue4u replied:
[e2e96c9](https://github.com/pixiv/three-vrm/pull/1427/commits/e2e96c9a27992cf7001ba81e27350643ae97dc7f)

0b5vr committed:
fix: MToonNodeMaterialParameters should be exported from `@pixiv/three-vrm-materials-mtoon/nodes`

0b5vr committed:
change: Change the MToonNodeMaterial API

Removed `MToonNodeMaterialLoaderPlugin` and introduce `materialType` property to `MToonMaterialLoaderPlugin`

Updated README, doc comments, and examples according to the new API

0b5vr committed:
Refactor: MToonMaterialLoaderPlugin, trivial refactor

move some of outline procedure to a separated function `_shouldGenerateOutline`

0b5vr committed:
feat: Add `isMToonNodeMaterial` to `MToonNodeMaterial` as other materials do

0b5vr committed:
fix: MToonNodeMaterial, remove `glEqualizationFactor`, `v0CompatShade`, `debugMode` from parameters

MToonNodeMaterial does not support these parameters so we delete them

0b5vr committed:
Merge pull request #1429 from pixiv/materialtype

feat: Change the MToonNodeMaterial API

yue4u committed:
fix: remove importPathPlugin since the bundling issue is lifted via reorganized deps graph in mtoon-loader

yue4u committed:
fix: seperate needsNodesBuild in bin/build.mjs for readability

Co-authored-by: 0b5vr <0b5vr@0b5vr.com>

yue4u committed:
fix: move addSuffix call into object def to align with extraEntryPoints

yue4u committed:
doc: update comment for addSuffix

Co-authored-by: 0b5vr <0b5vr@0b5vr.com>

yue4u commented:
webgpu-dnd example now produces a lot of errors when using with an order version of three. Can we do something about this?

![Screenshot 2024-07-01 at 15 35 05](https://github.com/pixiv/three-vrm/assets/26110087/59110640-2446-48e1-8990-69d523a9bda9)

0b5vr commented:
> webgpu-dnd example now produces a lot of errors when using with an order version of three. Can we do something about this?

`WebGPURenderer.renderAsync` has been introduced in r161. I believe it's fine in all versions of r161 or later

See: https://github.com/mrdoob/three.js/pull/27597

0b5vr committed:
feat: MToonNodeMaterial, Move pre-r161 warning to the top of the module

Errors can happen before instantiate MToonNodeMaterial
